### DISCUSSION TITLE
How do I delete RawContacts that have a specific Organization.company data?

### DISCUSSION THREAD

**Discussion by amoshalava** [vestrel00/contacts-android#157] [Q&A] [answered]
Awesome code by the way!!! 

Is there a way to delete a group based on the company name?

I'm writing an app that pulls company contacts from an API and want to delete all that has a #CompanyName and then replace with the new information from the API. I have it creating just fine but getting to delete is somehow not working for me.

**Comment by vestrel00** [comment]
Thanks for the compliment 😁 

Before I answer your question, I need to clarify something.

Are you trying to delete a set of `GroupEntity` OR a set of `ContactEntity` with a company name that matches what you want?

**Reply by amoshalava** [reply]
i'm not sure. i'm trying to figure out the way to do it. I have created a bunch of Contacts with the company set to #CompanyName. I just need to delete all of them before I add the list again. Is their a best practices way? should I group them someway first?

**Reply by amoshalava** [reply]
My intent is when someone leaves the company the contact is removed and the company phone number goes to a new person it would show up with the new person's name.

**Reply by vestrel00** [reply]
Ahh, I see! So what you want to do is first get the list of Contacts that have `Organization.company` matching the company name. Then, delete all of those contacts. Then, create new contacts from your API data.

If I understood that correctly, then this library can do exactly that 😁

I'll be able to give you the exact answer in code within the hour once I get to my laptop. Im currently on my phone lol

**Reply by amoshalava** [reply]
awesome thank you!!!

**Reply by vestrel00** [reply]
I updated the discussion title to reflect the question.

**Comment by vestrel00** [comment, marked as answer]
To answer your question effectively, it would be useful for you to read the **"Contacts Provider / ContactsContract Basic Concept"** section in this how-to/guide; https://github.com/vestrel00/contacts-android/blob/main/howto/howto-learn-more-about-api-entities.md

The rest of this answer assumes that you have a basic understanding of the one to many relationship between `Contacts` -> `RawContacts` -> `Data`.

I'll breakdown this answer using my favorite **problem** and **solution** style. Please let me know if I formulated the problem incorrectly. The answer will be incorrect if the problem is not accurate 😄 

## Problem

Given the following tables,

```
## Contacts
Contact_Id: 1, DisplayName: "Apple"
Contact_Id: 2, DisplayName: "Banana"
Contact_Id: 3, DisplayName: "Carrot"
Contact_Id: 4, DisplayName: "Durian"

## RawContacts
RawContact_Id: 1, Contact_Id: 1, DisplayName: "Apple"
RawContact_Id: 2, Contact_Id: 2, DisplayName: "Banana"
RawContact_Id: 3, Contact_Id: 3, DisplayName: "Carrot"
RawContact_Id: 4, Contact_Id: 3, DisplayName: "Calabash"
RawContact_Id: 5, Contact_Id: 4, DisplayName: "Durian"
RawContact_Id: 6, Contact_Id: 4, DisplayName: "Dragonfruit"

## Data
Data_Id: 1, RawContact_Id: 1, Organization.Company: "A Fruits"
Data_Id: 2, RawContact_Id: 2, Organization.Company: "B Fruits"
Data_Id: 3, RawContact_Id: 3, Organization.Company: "C Fruits"
Data_Id: 4, RawContact_Id: 4, Organization.Company: "C Fruits"
Data_Id: 5, RawContact_Id: 5, Organization.Company: "D Fruits"
Data_Id: 6, RawContact_Id: 6, Organization.Company: "Dragons"
```

Write functions to...

1. Delete all `Contact`s that have at least one `RawContact` that has `Organization.Company` equal to (case insensitive) `companyName`.
2. Delete all `RawContacts`s that have `Organization.Company` equal to (case insensitive) `companyName`.

These functions differ in behavior as will be discussed in the solution.

## Solution

#### (1) Delete all `Contact`s that have at least one `RawContact` that has `Organization.Company` equal to (case insensitive) `companyName`.

```kotlin
fun deleteAllContactsWithARawContactThatHasCompanyNameEqualTo(context: Context, companyName: String) {
    val contactsApi = Contacts(context)
    
    // First get the Contacts to delete.
    val contactsToDelete = contactsApi
        .query()
        .include(Fields.Contact.Id)
        .where { 
            Organization.Company equalToIgnoreCase companyName
        }
        .find()

    // Then delete the Contacts.
    contactsApi
        .delete()
        .contacts(contactsToDelete)
        .commit()
}
```

- If `companyName` is "a fruits", then...
    - Contacts with Id will be deleted; 1
    - also RawContacts with Id will be deleted; 1
- If `companyName` is "b fruits", then...
    - Contacts with Id will be deleted; 2
    - also RawContacts with Id will be deleted; 2
- If `companyName` is "c fruits", then...
    - Contacts with Id will be deleted; 3
    - also RawContacts with Id will be deleted; 3, 4
- If `companyName` is "d fruits", then...
    - Contacts with Id will be deleted; 4
    - also RawContacts with Id will be deleted; 5, 6
    - **Notice that RawContact with Id 6 still gets deleted even though its company name "Dragons" does not match "d fruits"!**

#### (2) Delete all `RawContacts`s that have `Organization.Company` equal to (case insensitive) `companyName`.

```kotlin
fun deleteAllRawContactsThatHasCompanyNameEqualTo(context: Context, companyName: String) {
    val contactsApi = Contacts(context)

    val rawContactsToDelete = contactsApi
        .query()
        .include(Fields.Organization.Company)
        .where { Organization.Company equalToIgnoreCase companyName }
        .find()
        .flatMap { it.rawContacts }
        .filter {
            it.organization?.company?.equals(companyName, ignoreCase = true) == true
        }

    contactsApi
        .delete()
        .rawContacts(rawContactsToDelete)
        .commit()
}
```

- If `companyName` is "a fruits", then...
    - Contacts with Id will be deleted; 1
    - also RawContacts with Id will be deleted; 1
- If `companyName` is "b fruits", then...
    - Contacts with Id will be deleted; 2
    - also RawContacts with Id will be deleted; 2
- If `companyName` is "c fruits", then...
    - Contacts with Id will be deleted; 3
    - also RawContacts with Id will be deleted; 3, 4
- If `companyName` is "d fruits", then...
    - **No Contacts will be deleted**
    - **Only RawContacts with Id 5 will be deleted**.
    - **Notice that RawContact with Id 6 DOES NOT get deleted because its company name "Dragons" does not match "d fruits"!**
    - **Contact with Id 4 DOES NOT get deleted because it still has one RawContact remaining".**
    - `Contact_Id: 4, DisplayName: "Durian"` transforms to `Contact_Id: 4, DisplayName: "Dragonfruit"` because the RawContact "Durian" was deleted but the RawContact "Dragonfruit" remains.

## Contacts vs RawContacts

Depending on the behavior that you want, you either want to delete `Contact`s (and all of its constituent `RawContact`s) OR just the `RawContact`s. Keep in mind that a `Contact` will NOT be deleted as long as it still has at least one `RawContact` remaining.

To help you see what I'm talking about, here is a recording of linking and unlinking Contacts (and their constituent RawContacts) using the Android Open Source Project [(AOSP) Contacts](https://play.google.com/store/apps/details?id=com.google.android.contacts).

https://user-images.githubusercontent.com/4332753/147841094-3a94aa3d-7551-49c5-9656-b79d1614eba4.mov

In the above video, you see that there are two Contacts each having one RawContact. Linking them results in a single Contact with two constituent RawContacts.

## Final thoughts

This library simply provides you a set of tools that greatly simplifies interfacing with the [Contacts Provider](https://developer.android.com/guide/topics/providers/contacts-provider). However, it does not define its own rules. It respects and follows the rules set forth by the Contacts Provider. This means that users of this library must understand the difference and relationship between `Contact`s and `RawContact`s. 

This is the reason why this library provides so much documentation in code and in the [howto pages](https://github.com/vestrel00/contacts-android/tree/main/howto). Any library that attempts to wrap the Contacts Provider MUST provide extensive documentation. Otherwise, users of the library will make mistakes. There is no way around this. 

What ever you can do with the AOSP Contacts app, which comes with almost every copy of Android, this library will allow you to do.

**Reply by vestrel00** [reply]
On a side note, I just created the following issue to make it possible to delete Contacts and RawContacts using a WHERE clause in a single query without having to fetch them first.

- #158

The new functionality will be a part of the [v0.1.11 Release](https://github.com/vestrel00/contacts-android/discussions/146).

**Reply by amoshalava** [reply]
This is great stuff! Keep it up. Thank you

**Reply by vestrel00** [reply]
Thanks. If this answered your question, please mark it as the answer.

Happy new year! 😁

**Comment by amoshalava** [comment]
![image](https://user-images.githubusercontent.com/15370732/147869284-d5abc7a4-09fa-4a45-b0bc-b0f7ad45f5da.png)
i keep getting this error when i paste that code into android studio.

**Reply by amoshalava** [reply]
![image](https://user-images.githubusercontent.com/15370732/147869316-93fd51dd-1b57-4d00-b431-5ded6a0ea995.png)
if i do it this way i don't get error but it doesn't delete the contacts.

**Reply by vestrel00** [reply]
> ![image](https://user-images.githubusercontent.com/15370732/147869284-d5abc7a4-09fa-4a45-b0bc-b0f7ad45f5da.png)
> 
> i keep getting this error when i paste that code into android studio.

Ohh, of course! Lol sorry. That syntax has not yet been released. It is merged in the main branch already though. I will be releasing v0.1.10 later this week.

I'm wondering what the issue is. In my examples above, are you getting non-empty lists for `contactsToDelete` and/or `rawContactsToDelete`? If they are empty, then that means that the query was not able to match any contacts.

If the query results are not empty, then the issue might be with the delete operation. Please make sure you have `Manifest.permission.WRITE_CONTACTS` granted. Also, let me know the output of the delete `Result.isSuccessful`.

Also, please use the default contacts app to make sure that the contacts you want to delete have the organization company you want to match. Beware that organization company and title are different so maybe that might be getting mixed up too?

I will check tomorrow morning to see if I can reproduce the issue you might be seeing.

Ohh, one more thing I want to clarify. Is the issue that the contacts are not being deleted locally on the device? Or is it that it is not being deleted across devices and in web immediately?

**Reply by amoshalava** [reply]
with this release the code you sent me should work now right?

**Reply by vestrel00** [reply]
The compile error you were seeing with the syntax will go away. However, I don't know if the issue you are seeing is fixed. I asked you for more details in my previous comment. If you could answer those questions for me, I will be able to help you better =)

**Reply by amoshalava** [reply]
ok i'm gonna try figure this out how to update and see if it work. i thought i had to wait till you updated it. i'll work on it right now but may have to finish tomorrow. this is great.

**Reply by amoshalava** [reply]
i'm only writing my contacts to the device i don't want them synced across the web i want the app to sync them.

**Reply by vestrel00** [reply]
Thanks! Sorry for the misunderstanding! Looking forward to your investigation results =)

**Reply by amoshalava** [reply]
i call the code like this:

                    deleteAllRawContactsThatHasCompanyNameEqualTo(this, "#CCSO AgencySync")
                    deleteAllContactsWithARawContactThatHasCompanyNameEqualTo(this, "#CCSO AgencySync")

**Reply by amoshalava** [reply]
private fun deleteAllRawContactsThatHasCompanyNameEqualTo(context: Context, companyName: String) {
        val contactsApi = Contacts(context)

        val rawContactsToDelete = contactsApi
            .query()
            .include(Fields.Organization.Company)
            .where { Organization.Company equalToIgnoreCase companyName }
            //.where( Organization.Company equalTo companyName)
            .find()
            .flatMap { it.rawContacts }
            .filter {
                it.organization?.company?.equals(companyName, ignoreCase = true) == true
            }

        contactsApi
            .delete()
            .rawContacts(rawContactsToDelete)
            .commit()
    }

    private fun deleteAllContactsWithARawContactThatHasCompanyNameEqualTo(context: Context, companyName: String) {
        val contactsApi = Contacts(context)

        // First get the Contacts to delete.
        val contactsToDelete = contactsApi
            .query()
            .include(Fields.Contact.Id)
            //.where( Organization.Company equalTo companyName)
            .where {
                Organization.Company equalToIgnoreCase companyName
            }
            .find()

        // Then delete the Contacts.
        contactsApi
            .delete()
            .contacts(contactsToDelete)
            .commit()
    }

**Reply by vestrel00** [reply]
> i'm only writing my contacts to the device i don't want them synced across the web i want the app to sync them.

Contact syncing behavior is handled by the Android Contacts Provider automatically depending on Account sync settings (or lack thereof). Please read these howto pages to learn more about contact syncing;

- https://github.com/vestrel00/contacts-android/blob/main/howto/howto-sync-contact-data.md
- https://github.com/vestrel00/contacts-android/blob/main/howto/howto-learn-more-about-local-contacts.md

**Reply by amoshalava** [reply]
using this code. but they don't delete. i'm using the device contacts. it adds them ok. if i rename just the name on one it will add another one with that name just fine but it doesn't delete anything.

**Reply by amoshalava** [reply]
So the app design is to pull Company phone numbers from an API, create contacts on just the device so they don't sync across to any accounts on the phone, then when someone phone changes or someone leaves company and a new person gets their number it will delete all the contacts that I previously added and then add all new ones to that the names update to new numbers.

**Reply by vestrel00** [reply]
>  create contacts on just the device so they don't sync across to any accounts on the phone

You can do this. However, there are some caveats you need to know. Everything should be documented in the function documentation of `Insert.forAccount`.

<img width="650" alt="Screen Shot 2022-01-07 at 5 44 16 PM" src="https://user-images.githubusercontent.com/4332753/148630387-ff1fdb39-f797-43a8-ad79-adb0e6ec119b.png">

**Reply by vestrel00** [reply]
> i call the code like this:

The code you have seems fine. Please do this for me,

1. Upgrade to 0.1.10 (I think you already did but just making sure).
2. In all of your code, instead of `Contacts(context)`, do `Contacts(context, logger = AndroidLogger(redactMessages = false))`
3. Download the [Google Contacts app](https://play.google.com/store/apps/details?id=com.google.android.contacts&hl=en_IN&gl=US)
4. Create some fake contacts with company name "#CCSO AgencySync" using the **Google Contacts app**.
5. Then, run your app  with the function `deleteAllContactsWithARawContactThatHasCompanyNameEqualTo(this, "#CCSO AgencySync")`
6. Filter your Logcat with the tag "ContactsDebug".
7. Copy-paste the logs in this discussion so I can see it.
    - **Please make sure to not copy-paste private information! Only copy-paste logs from the fake contacts you created!**

**Reply by amoshalava** [reply]
OH MAN IT WORKS!!! I'm so sorry i had to reset permissions. I'll write code to prompt for permission so i don't miss that again. this is huge. let me know how I can help promote this!

**Reply by vestrel00** [reply]
I'm so glad it works for you!!! I'm just happy that I was able to help. There is no need for you to promote this but if you really want to, then feel free to let people know about this library =)

One thing you can do is ⭐️ star the repo if you haven't already. If you want to get updates, also 👁 watch the repo.

Cheers 🍻
